File: backend/search/search_queries.py
import mysql.connector
import re
from difflib import SequenceMatcher

# Database configuration
DB_CONFIG = {
    'host': 'localhost',
    'user': 'root',  # Change this to your MySQL username
    'password': '',  # Change this to your MySQL password
    'database': 'person_search_db'
}

class PersonSearch:
    def __init__(self):
        self.connection = None
        self.cursor = None
        
    def connect(self):
        """Establish database connection"""
        try:
            self.connection = mysql.connector.connect(**DB_CONFIG)
            self.cursor = self.connection.cursor(dictionary=True)
            return True
        except mysql.connector.Error as e:
            print(f"Error connecting to MySQL: {e}")
            return False
    
    def close(self):
        """Close database connection"""
        if self.cursor:
            self.cursor.close()
        if self.connection:
            self.connection.close()
    
    def normalize_string(self, s):
        """Normalize string for comparison"""
        # Remove spaces, apostrophes, and convert to lowercase
        return re.sub(r'[^a-zA-Z0-9]', '', s).lower()
    
    def exact_search(self, search_term):
        """Exact match search"""
        query = """
        SELECT * FROM persons 
        WHERE first_name = %s 
           OR last_name = %s 
           OR full_name = %s
        LIMIT 10
        """
        self.cursor.execute(query, (search_term, search_term, search_term))
        return self.cursor.fetchall()
    
    def wildcard_search(self, search_term):
        """Search using SQL wildcards"""
        wildcard_term = f"%{search_term}%"
        query = """
        SELECT * FROM persons 
        WHERE first_name LIKE %s 
           OR last_name LIKE %s 
           OR full_name LIKE %s
        LIMIT 20
        """
        self.cursor.execute(query, (wildcard_term, wildcard_term, wildcard_term))
        return self.cursor.fetchall()
    
    def normalized_search(self, search_term):
        """Search after normalizing strings (removing spaces, apostrophes)"""
        normalized = self.normalize_string(search_term)
        query = """
        SELECT * FROM persons 
        WHERE REPLACE(REPLACE(REPLACE(LOWER(first_name), ' ', ''), "'", ''), '-', '') = %s
           OR REPLACE(REPLACE(REPLACE(LOWER(last_name), ' ', ''), "'", ''), '-', '') = %s
           OR REPLACE(REPLACE(REPLACE(LOWER(full_name), ' ', ''), "'", ''), '-', '') = %s
        LIMIT 20
        """
        self.cursor.execute(query, (normalized, normalized, normalized))
        return self.cursor.fetchall()
    
    def fulltext_search(self, search_term):
        """MySQL Full-text search"""
        query = """
        SELECT *, MATCH(first_name, last_name, full_name) AGAINST(%s) as relevance
        FROM persons 
        WHERE MATCH(first_name, last_name, full_name) AGAINST(%s)
        ORDER BY relevance DESC
        LIMIT 20
        """
        self.cursor.execute(query, (search_term, search_term))
        return self.cursor.fetchall()
    
    def soundex_search(self, search_term):
        """Search using SOUNDEX for phonetic matching"""
        query = """
        SELECT * FROM persons 
        WHERE SOUNDEX(first_name) = SOUNDEX(%s)
           OR SOUNDEX(last_name) = SOUNDEX(%s)
           OR SOUNDEX(CONCAT(first_name, last_name)) = SOUNDEX(REPLACE(%s, ' ', ''))
        LIMIT 20
        """
        self.cursor.execute(query, (search_term, search_term, search_term))
        return self.cursor.fetchall()
    
    def levenshtein_search(self, search_term, max_distance=3):
        """Search using Levenshtein distance"""
        # First get a subset of candidates using wildcard search
        first_chars = search_term[:3] if len(search_term) >= 3 else search_term
        query = """
        SELECT * FROM persons 
        WHERE first_name LIKE %s 
           OR last_name LIKE %s 
           OR full_name LIKE %s
        """
        self.cursor.execute(query, (f"{first_chars}%", f"{first_chars}%", f"{first_chars}%"))
        candidates = self.cursor.fetchall()
        
        # Calculate Levenshtein distance for each candidate
        results = []
        for candidate in candidates:
            distances = []
            
            # Check against different fields
            for field in ['first_name', 'last_name', 'full_name']:
                if candidate[field]:
                    # Calculate distance using MySQL function
                    dist_query = "SELECT LEVENSHTEIN(%s, %s) as distance"
                    self.cursor.execute(dist_query, (search_term.lower(), candidate[field].lower()))
                    distance = self.cursor.fetchone()['distance']
                    distances.append(distance)
            
            # Use minimum distance
            if distances and min(distances) <= max_distance:
                candidate['levenshtein_distance'] = min(distances)
                results.append(candidate)
        
        # Sort by distance
        results.sort(key=lambda x: x['levenshtein_distance'])
        return results[:20]
    
    def combined_search(self, search_term):
        """Combine multiple search strategies"""
        results = {}
        scores = {}
        
        # 1. Exact search (highest priority)
        exact_results = self.exact_search(search_term)
        for r in exact_results:
            key = r['id']
            results[key] = r
            scores[key] = scores.get(key, 0) + 100
        
        # 2. Normalized search (high priority for handling spaces/apostrophes)
        normalized_results = self.normalized_search(search_term)
        for r in normalized_results:
            key = r['id']
            if key not in results:
                results[key] = r
            scores[key] = scores.get(key, 0) + 80
        
        # 3. Soundex search (medium priority)
        soundex_results = self.soundex_search(search_term)
        for r in soundex_results:
            key = r['id']
            if key not in results:
                results[key] = r
            scores[key] = scores.get(key, 0) + 60
        
        # 4. Wildcard search (low priority)
        wildcard_results = self.wildcard_search(search_term)
        for r in wildcard_results:
            key = r['id']
            if key not in results:
                results[key] = r
            scores[key] = scores.get(key, 0) + 40
        
        # 5. Levenshtein search for remaining fuzzy matches
        levenshtein_results = self.levenshtein_search(search_term, max_distance=3)
        for r in levenshtein_results:
            key = r['id']
            if key not in results:
                results[key] = r
            # Score inversely proportional to distance
            distance = r.get('levenshtein_distance', 3)
            scores[key] = scores.get(key, 0) + (30 / (distance + 1))
        
        # Sort by score
        sorted_results = sorted(results.items(), key=lambda x: scores[x[0]], reverse=True)
        return [r[1] for r in sorted_results[:10]]
    
    def display_results(self, results, search_term, method):
        """Display search results in a formatted way"""
        print(f"\n{'='*80}")
        print(f"Search Method: {method}")
        print(f"Search Term: '{search_term}'")
        print(f"Results Found: {len(results)}")
        print(f"{'='*80}")
        
        if not results:
            print("No results found.")
            return
        
        print(f"\n{'ID':<6} {'First Name':<20} {'Last Name':<20} {'Full Name':<30}")
        print(f"{'-'*6} {'-'*20} {'-'*20} {'-'*30}")
        
        for r in results[:10]:  # Show max 10 results
            print(f"{r['id']:<6} {r['first_name']:<20} {r['last_name']:<20} {r['full_name']:<30}")

def run_test_searches():
    """Run test searches with the examples provided"""
    search = PersonSearch()
    
    if not search.connect():
        print("Failed to connect to database")
        return
    
    test_cases = [
        ("MaryAnn", "Mary Ann"),
        ("Mary Ann", "MaryAnn"),
        ("D'Souza", "DSouza"),
        ("DSouza", "D'Souza"),
        ("Mirage Air Craft", "Mirage AirCraft"),
        ("Mirage AirCraft", "Mirage Air Craft"),
        ("Jean-Pierre", "JeanPierre"),
        ("O'Brien", "OBrien"),
        ("MacDonald", "Mac Donald"),
    ]
    
    print("\n" + "="*80)
    print("RUNNING TEST SEARCHES")
    print("="*80)
    
    for search_term, expected in test_cases:
        print(f"\n\n{'*'*80}")
        print(f"SEARCHING FOR: '{search_term}' (expecting to find variations like '{expected}')")
        print(f"{'*'*80}")
        
        # Run combined search (best approach)
        results = search.combined_search(search_term)
        search.display_results(results, search_term, "COMBINED SEARCH (Best Results)")
        
        # Also show individual search methods for comparison
        print("\n--- Individual Search Methods for Comparison ---")
        
        # Normalized search (good for spaces/apostrophes)
        norm_results = search.normalized_search(search_term)
        if norm_results:
            print(f"\nNormalized Search found {len(norm_results)} results")
            
        # Soundex search
        soundex_results = search.soundex_search(search_term)
        if soundex_results:
            print(f"Soundex Search found {len(soundex_results)} results")
            
        # Levenshtein search
        lev_results = search.levenshtein_search(search_term)
        if lev_results:
            print(f"Levenshtein Search found {len(lev_results)} results")
    
    search.close()

def interactive_search():
    """Run interactive search"""
    search = PersonSearch()
    
    if not search.connect():
        print("Failed to connect to database")
        return
    
    print("\n" + "="*80)
    print("INTERACTIVE PERSON SEARCH")
    print("="*80)
    print("Type 'quit' to exit")
    
    while True:
        search_term = input("\nEnter search term: ").strip()
        
        if search_term.lower() == 'quit':
            break
        
        if not search_term:
            continue
        
        # Use combined search for best results
        results = search.combined_search(search_term)
        search.display_results(results, search_term, "COMBINED SEARCH")
    
    search.close()
    print("\nGoodbye!")

if __name__ == "__main__":
    # Run test searches
    run_test_searches()
    
    # Run interactive search
    print("\n\n")
    interactive_search() 

File: backend/search/search_queries.ts
import mysql from 'mysql2/promise';
import * as readline from 'readline';
import { DB_CONFIG, Person } from './db_config';

interface SearchResult extends Person {
  score?: number;
  foundBy?: string[];
}

export default class PersonSearch {
  connection: mysql.Connection | null = null;

  async connect(): Promise<boolean> {
    try {
      this.connection = await mysql.createConnection(DB_CONFIG);
      return true;
    } catch (error) {
      console.error('Error connecting to MySQL:', error);
      return false;
    }
  }

  async close(): Promise<void> {
    if (this.connection) {
      await this.connection.end();
    }
  }

  normalizeString(s: string): string {
    // Only remove spaces and convert to lowercase, keep apostrophes
    return s.replace(/\s+/g, '').toLowerCase();
  }

  async exactSearch(searchTerm: string): Promise<Person[]> {
    const query = `
      SELECT * FROM persons 
      WHERE full_name = ?
      LIMIT 10
    `;
    
    const [rows] = await this.connection!.execute(query, [searchTerm]);
    return rows as Person[];
  }

  async wildcardSearch(searchTerm: string): Promise<Person[]> {
    // Handle both versions of the name (with and without apostrophe)
    const normalizedTerm = searchTerm.replace("'", "");
    const query = `
      SELECT * FROM persons 
      WHERE LOWER(full_name) LIKE LOWER(?) 
         OR LOWER(full_name) LIKE LOWER(?)
      LIMIT 20
    `;
    
    const [rows] = await this.connection!.execute(query, [`%${searchTerm}%`, `%${normalizedTerm}%`]);
    return rows as Person[];
  }

  async normalizedSearch(searchTerm: string): Promise<Person[]> {
    // Split search term into words and create LIKE conditions
    const searchWords = searchTerm.toLowerCase().split(/\s+/).filter(w => w.length > 0);
    const likeConditions = searchWords.map(word => `LOWER(full_name) LIKE ?`).join(' AND ');
    
    const query = `
      SELECT * FROM persons 
      WHERE ${likeConditions}
      LIMIT 20
    `;
    
    // Create parameters array with %word% for each search word
    const params = searchWords.map(word => `%${word}%`);
    
    const [rows] = await this.connection!.execute(query, params);
    return rows as Person[];
  }

  async fulltextSearch(searchTerm: string): Promise<Person[]> {
    // Still using the existing fulltext index on all fields, but only checking full_name
    const query = `
      SELECT *, MATCH(first_name, last_name, full_name) AGAINST(?) as relevance
      FROM persons 
      WHERE MATCH(first_name, last_name, full_name) AGAINST(?)
        AND full_name LIKE CONCAT('%', ?, '%')
      ORDER BY relevance DESC
      LIMIT 20
    `;
    
    const [rows] = await this.connection!.execute(query, [searchTerm, searchTerm, searchTerm]);
    return rows as Person[];
  }

  async soundexSearch(searchTerm: string): Promise<Person[]> {
    // More flexible: handle phonetically similar first letters
    const firstChar = searchTerm.substring(0, 1).toLowerCase();
    const firstTwoChars = searchTerm.substring(0, 2).toLowerCase();
    
    // Define phonetically equivalent letters
    const phoneticGroups: { [key: string]: string[] } = {
      'c': ['c', 'k', 's'],
      'k': ['k', 'c'],
      's': ['s', 'c', 'z'],
      'z': ['z', 's'],
      'f': ['f', 'ph'],
      'j': ['j', 'g'],
      'g': ['g', 'j'],
      'x': ['x', 'z'],
      'q': ['q', 'k'],
      'v': ['v', 'w'],
      'w': ['w', 'v']
    };
    
    // Get phonetic equivalents for the first character
    // const equivalentChars = phoneticGroups[firstChar] || [firstChar];
    const equivalentChars = [firstChar];
    const firstCharConditions = equivalentChars.map(char => 
      `(LOWER(SUBSTRING(full_name, 1, 1)) = ? AND SUBSTRING(SOUNDEX(full_name), 2) = SUBSTRING(SOUNDEX(?), 2))`
    ).join(' OR ');
    
    const query = `
      SELECT * FROM persons 
      WHERE ${firstCharConditions}
         OR (SOUNDEX(full_name) = SOUNDEX(?) AND LOWER(SUBSTRING(full_name, 1, 2)) = ?)
      LIMIT 20
    `;
    
    // Build parameters array
    const params: string[] = [];
    equivalentChars.forEach(char => {
      params.push(char, searchTerm);
    });
    params.push(searchTerm, firstTwoChars);
    
    const [rows] = await this.connection!.execute(query, params);
    return rows as Person[];
  }

  async levenshteinSearch(searchTerm: string, maxDistance: number = 2): Promise<Person[]> {
    // More strict: reduced from 3 to 2, and added percentage threshold
    const firstChars = searchTerm.length >= 3 ? searchTerm.substring(0, 3) : searchTerm;
    const query = `
      SELECT * FROM persons 
      WHERE full_name LIKE ?
    `;
    
    const [candidates] = await this.connection!.execute(query, [`${firstChars}%`]);
    
    // Calculate Levenshtein distance for each candidate
    const results: Person[] = [];
    
    for (const candidate of candidates as Person[]) {
      if (candidate.full_name) {
        // Calculate distance using MySQL function
        const distQuery = "SELECT LEVENSHTEIN(?, ?) as distance";
        const [distResult] = await this.connection!.execute(distQuery, 
          [searchTerm.toLowerCase(), candidate.full_name.toLowerCase()]);
        const distance = (distResult as any)[0].distance;
        
        // Calculate percentage threshold - distance should be less than 30% of the longer string
        const maxLength = Math.max(searchTerm.length, candidate.full_name.length);
        const percentageThreshold = Math.ceil(maxLength * 0.3);
        
        if (distance <= Math.min(maxDistance, percentageThreshold)) {
          candidate.levenshtein_distance = distance;
          results.push(candidate);
        }
      }
    }
    
    // Sort by distance
    results.sort((a, b) => (a.levenshtein_distance || 0) - (b.levenshtein_distance || 0));
    return results.slice(0, 20);
  }

  async strictLevenshteinSearch(searchTerm: string, maxDistance: number = 3): Promise<Person[]> {
    const query = `
      SELECT *, LEVENSHTEIN(full_name, ?) AS distance
      FROM persons
      HAVING distance <= ?
      ORDER BY distance ASC
      LIMIT 20
    `;
    const [rows] = await this.connection!.execute(query, [
      searchTerm.toLowerCase(),
      Math.min(maxDistance, Math.ceil(searchTerm.length * 0.3)),
    ]);
    return rows as Person[];
  }

  async strictSoundexSearch(searchTerm: string, maxDistance: number = 3): Promise<Person[]> {
    const query = `
      SELECT *, LEVENSHTEIN(full_name, ?) AS distance
      FROM persons 
      WHERE SOUNDEX(full_name) = SOUNDEX(?)
      HAVING distance <= ?
      ORDER BY distance ASC
      LIMIT 20
    `; // distance should be dynamic based on the search term length
    const [rows] = await this.connection!.execute(query, [searchTerm, searchTerm, Math.min(maxDistance, Math.ceil(searchTerm.length * 0.3)),    ]);
    return rows as Person[];
  }

  async combinedSearch(searchTerm: string): Promise<SearchResult[]> {
    const results: Map<number, SearchResult> = new Map();
    const scores: Map<number, number> = new Map();
    const foundBy: Map<number, Set<string>> = new Map();
    
    // 1. Exact search (highest priority)
    const exactResults = await this.exactSearch(searchTerm);
    for (const r of exactResults) {
      results.set(r.id, r);
      scores.set(r.id, (scores.get(r.id) || 0) + 10);
      if (!foundBy.has(r.id)) foundBy.set(r.id, new Set());
      foundBy.get(r.id)!.add('Exact');
    }
    
    // 2. Normalized search (high priority for handling spaces/apostrophes)
    const normalizedResults = await this.normalizedSearch(searchTerm);
    for (const r of normalizedResults) {
      if (!results.has(r.id)) {
        results.set(r.id, r);
      }
      scores.set(r.id, (scores.get(r.id) || 0) + 8);
      if (!foundBy.has(r.id)) foundBy.set(r.id, new Set());
      foundBy.get(r.id)!.add('Normalized');
    }
    
    // 3. Strict Soundex search (high priority - exact soundex match)
    const strictSoundexResults = await this.strictSoundexSearch(searchTerm, searchTerm.replace(/\s/g,'').length * 0.6);
    for (const r of strictSoundexResults) {
      if (!results.has(r.id)) {
        results.set(r.id, r);
      }
      scores.set(r.id, (scores.get(r.id) || 0) + 4);
      if (!foundBy.has(r.id)) foundBy.set(r.id, new Set());
      foundBy.get(r.id)!.add('Soundex');
    }
    
    // 5. Wildcard search (low priority)
    const wildcardResults = await this.wildcardSearch(searchTerm);
    for (const r of wildcardResults) {
      if (!results.has(r.id)) {
        results.set(r.id, r);
      }
      scores.set(r.id, (scores.get(r.id) || 0) + 2);
      if (!foundBy.has(r.id)) foundBy.set(r.id, new Set());
      foundBy.get(r.id)!.add('Wildcard');
    }
    
    const strictLevenshteinResults = await this.strictLevenshteinSearch(searchTerm, searchTerm.replace(/\s/g,'').length * 0.6);
    for (const r of strictLevenshteinResults) {
      if (!results.has(r.id)) {
        results.set(r.id, r);
      }
      // Increased base score since it's more strict now
      scores.set(r.id, (scores.get(r.id) || 0) + 6);
      if (!foundBy.has(r.id)) foundBy.set(r.id, new Set());
      foundBy.get(r.id)!.add(`Levenshtein`);
    }
    
    // Sort by score
    const sortedResults = Array.from(results.entries())
      .map(([id, person]) => ({ 
        ...person, 
        score: scores.get(id) || 0,
        foundBy: Array.from(foundBy.get(id) || [])
      }))
      .sort((a, b) => b.score - a.score)
      .slice(0, 10);
    
    return sortedResults;
  }

  displayResults(results: SearchResult[], searchTerm: string, method: string): void {
    console.log('\n' + '='.repeat(80));
    console.log(`Search Method: ${method}`);
    console.log(`Search Term: '${searchTerm}'`);
    console.log(`Results Found: ${results.length}`);
    console.log('='.repeat(80));
    
    if (results.length === 0) {
      console.log('No results found.');
      return;
    }
    
    console.log(`\n${'ID'.padEnd(6)} ${'Full Name'.padEnd(40)} ${'Score'.padEnd(10)} ${'Found By'.padEnd(30)}`);
    console.log(`${'-'.repeat(6)} ${'-'.repeat(40)} ${'-'.repeat(10)} ${'-'.repeat(30)}`);
    
    for (const r of results.slice(0, 10)) {
      const foundByStr = r.foundBy ? r.foundBy.join(', ') : '';
      const score = r.score ? r.score.toFixed(2) : '0.00';
      console.log(`${r.id.toString().padEnd(6)} ${r.full_name.padEnd(40)} ${score.padEnd(10)} ${foundByStr.padEnd(30)}`);
    }
  }

  async debugSoundex(searchTerm: string, compareTerms: string[]): Promise<void> {
    console.log('\n--- Soundex Debug ---');
    console.log(`Search term: "${searchTerm}"`);
    
    // Get soundex of search term
    const [searchSoundex] = await this.connection!.execute(
      'SELECT SOUNDEX(?) as soundex', 
      [searchTerm]
    );
    console.log(`Soundex of "${searchTerm}": ${(searchSoundex as any)[0].soundex}`);
    
    console.log('\nComparing with:');
    for (const term of compareTerms) {
      const [termSoundex] = await this.connection!.execute(
        'SELECT SOUNDEX(?) as soundex', 
        [term]
      );
      console.log(`  "${term}": ${(termSoundex as any)[0].soundex}`);
    }
  }
}

async function runTestSearches(): Promise<void> {
  const search = new PersonSearch();
  
  if (!await search.connect()) {
    console.log('Failed to connect to database');
    return;
  }
  
  const testCases: [string, string][] = [
    ["MaryAnn", "Mary Ann"],
    ["Mary Ann", "MaryAnn"],
    ["D'Souza", "DSouza"],
    ["DSouza", "D'Souza"],
    ["Mirage Air Craft", "Mirage AirCraft"],
    ["Mirage AirCraft", "Mirage Air Craft"],
    ["Jean-Pierre", "JeanPierre"],
    ["O'Brien", "OBrien"],
    ["MacDonald", "Mac Donald"],
  ];
  
  console.log('\n' + '='.repeat(80));
  console.log('RUNNING TEST SEARCHES');
  console.log('='.repeat(80));
  
  for (const [searchTerm, expected] of testCases) {
    console.log('\n\n' + '*'.repeat(80));
    console.log(`SEARCHING FOR: '${searchTerm}' (expecting to find variations like '${expected}')`);
    console.log('*'.repeat(80));
    
    // Run combined search (best approach)
    const results = await search.combinedSearch(searchTerm);
    search.displayResults(results, searchTerm, "COMBINED SEARCH (Best Results)");
    
    // Also show individual search methods for comparison
    console.log('\n--- Individual Search Methods for Comparison ---');
    
    // Normalized search (good for spaces/apostrophes)
    const normResults = await search.normalizedSearch(searchTerm);
    if (normResults.length > 0) {
      console.log(`\nNormalized Search found ${normResults.length} results`);
    }
    
    // Soundex search
    const soundexResults = await search.soundexSearch(searchTerm);
    if (soundexResults.length > 0) {
      console.log(`Soundex Search found ${soundexResults.length} results`);
    }
    
    // Levenshtein search
    const levResults = await search.levenshteinSearch(searchTerm);
    if (levResults.length > 0) {
      console.log(`Levenshtein Search found ${levResults.length} results`);
    }
  }
  
  await search.close();
}

async function interactiveSearch(): Promise<void> {
  const search = new PersonSearch();
  
  if (!await search.connect()) {
    console.log('Failed to connect to database');
    return;
  }
  
  console.log('\n' + '='.repeat(80));
  console.log('INTERACTIVE PERSON SEARCH');
  console.log('='.repeat(80));
  console.log("Type 'quit' to exit");
  console.log("Type 'debug soundex <term1> <term2> ...' to debug soundex values");
  
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });
  
  const askQuestion = (): Promise<void> => {
    return new Promise((resolve) => {
      rl.question('\nEnter search term: ', async (input) => {
        const trimmedInput = input.trim();
        
        if (trimmedInput.toLowerCase() === 'quit') {
          rl.close();
          await search.close();
          console.log('\nGoodbye!');
          resolve();
          return;
        }
        
        // Check for debug command
        if (trimmedInput.toLowerCase().startsWith('debug soundex ')) {
          const terms = trimmedInput.substring(14).split(' ').filter(t => t.length > 0);
          if (terms.length > 0) {
            const [searchTerm, ...compareTerms] = terms;
            await search.debugSoundex(searchTerm, compareTerms);
          }
        } else if (trimmedInput) {
          // Use combined search for best results
          const results = await search.combinedSearch(trimmedInput);
          search.displayResults(results, trimmedInput, "COMBINED SEARCH");
        }
        
        // Continue asking
        resolve(askQuestion());
      });
    });
  };
  
  await askQuestion();
}

async function main(): Promise<void> {
  // Run test searches
  // await runTestSearches();
  
  // Run interactive search
  console.log('\n\n');
  await interactiveSearch();
}

// Run if this is the main module
if (require.main === module) {
  main().catch(console.error);
} 